.github/workflows/inspektor-gadget.yml

```
@@ -195,25 +210,105 @@ jobs:
         container-image-dev: ${{ env.DEVELOPMENT_CONTAINER_REPO }}
-    - name: Build gadget ${{ matrix.type }} container
+    - name: Build gadget ${{ matrix.type }} ${{ matrix.os }} ${{ matrix.platform }} container image as artifacts
       uses: docker/build-push-action@v2
       with:
         context: /home/runner/work/inspektor-gadget/inspektor-gadget/
         file: /home/runner/work/inspektor-gadget/inspektor-gadget/Dockerfiles/gadget-${{ matrix.type }}.Dockerfile
         build-args: |
           ENABLE_BTFGEN=true
-        # At the moment, we only build core image and do not publish them.
-        push: ${{ matrix.type != 'core' }}
+        outputs: type=docker,dest=/tmp/gadget-container-image-${{ matrix.type }}-${{ matrix.os }}-${{ matrix.platform }}.tar
         tags: ${{ steps.choose-repo-determine-image-tag.outputs.container-repo }}:${{ steps.choose-repo-determine-image-tag.outputs.image-tag }}
         cache-from: type=local,src=/tmp/.buildx-cache
-        cache-to: type=local,dest=/tmp/.buildx-cache-new
-        # For the moment, we only support these two platforms.
-        platforms: linux/amd64,linux/arm64
+        cache-to: type=local,dest=/tmp/.buildx-cache-docker
+        # docker exporter doesn't allow exporting multi-platform images
+        platforms: ${{ matrix.os }}/${{ matrix.platform }}
+
+    - name: Publish gadget ${{ matrix.type }} ${{ matrix.os }} ${{ matrix.platform }} container image as artifacts
+      uses: actions/upload-artifact@master
+      with:
+        name: gadget-container-image-${{ matrix.type }}-${{ matrix.os }}-${{ matrix.platform }}.tar
+        path: /tmp/gadget-container-image-${{ matrix.type }}-${{ matrix.os }}-${{ matrix.platform }}.tar
+        retention-days: 1
+
+    # build time will not be increased with this workflow because of internal cache
+    # buildx is used here since it allows push-by-digest to avoid platform specific tags
+    - name: Publish gadget ${{ matrix.type }} ${{ matrix.os }} ${{ matrix.platform }} container image to registry
+      id: publish-gadget-container-images
+      if: github.event_name != 'pull_request'
+      uses: docker/build-push-action@v2
+      with:
+        context: /home/runner/work/inspektor-gadget/inspektor-gadget/
+        file: /home/runner/work/inspektor-gadget/inspektor-gadget/Dockerfiles/gadget-${{ matrix.type }}.Dockerfile
+        build-args: |
+          ENABLE_BTFGEN=true
+        outputs: type=registry,push=true,push-by-digest=true
+        tags: ${{ steps.choose-repo-determine-image-tag.outputs.container-repo }}
+        cache-from: type=local,src=/tmp/.buildx-cache-docker
+        cache-to: type=local,dest=/tmp/.buildx-cache-registry
+        # docker exporter doesn't allow exporting multi-platform images
+        platforms: ${{ matrix.os }}/${{ matrix.platform }}
+
+    - name: Prepare gadget ${{ matrix.type }} ${{ matrix.os }} ${{ matrix.platform }} container image digest
```

Maybe the following is more clear?

```suggestion
    - name: Save gadget ${{ matrix.type }} ${{ matrix.os }} ${{ matrix.platform }} container image digest output
```